### PR TITLE
Add REST backend option to use CA root certificate

### LINKED ISSUE #1114: Restic doesn't accept self-signed certificates
## Output of `restic version`

restic 0.7.0
compiled with go1.8.3 on linux/amd64

## How did you start restic exactly? (Include the complete command line)

restic init --repo "rest:https://backup:backup@192.168.1.250:8888/repo_name"

## What backend/server/service did you use?

Caddy's plugin

## Expected behavior

Repo inited and I can do backups over HTTPS

## Actual behavior

create backend at rest:https://backup:backup@192.168.1.250:8888/repo_name failed: Post https://backup:backup@192.168.1.250:8888/repo_name?create=true: x509: certificate signed by unknown authority

There is no option to accept self-signed certificates. I can't easily issue LE certificate because this VM isolated from WAN.

## Steps to reproduce the behavior

Create Caddyfile with HTTP and HTTPS (or just HTTP) enabled, specify restic's plugin options, start Caddy. Issue this command on client:

restic init --repo "rest:https://backup:backup@192.168.1.250:8888/repo_name"

### PR DESCRIPTION
**Pull request by fawick**
Fixes #1114.

### CODE REVIEW

**Comment by fd0** [comment]
Hm, it feels to me that the REST backend is the wrong place to add a CA certificate. What about moving the code to https://github.com/restic/restic/blob/49d397a4190de017ace5802962aeda22c518c67c/internal/backend/http_transport.go#L12?

**Comment by fawick** [comment]
Fair enough. Will refactor.

**Comment by fawick** [comment]
So, I moved that part to package `internal/backend`. The only one using it so far is still the REST backend, as this was specifically was #1114 was about. It might be useful for the S3 backend (when Minio is used). But probably not for GCS / Azure / Swift, I guess. What do you think, @fd0?

File: internal/backend/http_transport.go [hunk 1]
```diff
@@ -1,15 +1,25 @@
...
 	"github.com/restic/restic/internal/debug"
 )
 
-// Transport returns a new http.RoundTripper with default settings applied.
-func Transport() http.RoundTripper {
+// Transport returns a new http.RoundTripper with default settings applied. If
+// a custom rootCertFilename is non-empty, it must point to a valid PEM file,
+// otherwise the function will return an error.
+func Transport(rootCertFilename string) (http.RoundTripper, error) {
+	p, err := certPoolFromFile(rootCertFilename)
```
Comment (fd0) [inline internal/backend/http_transport.go]: This needs to be guarded with an `if rootCertFilename == ""`.

File: internal/backend/http_transport.go [hunk 2]
```diff
@@ -23,8 +33,23 @@ func Transport() http.RoundTripper {
 		IdleConnTimeout:       90 * time.Second,
 		TLSHandshakeTimeout:   10 * time.Second,
 		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig: &tls.Config{
```
Comment (fd0) [inline internal/backend/http_transport.go]: Same here: Only set this if a filename is present.

**Review by fd0 — CHANGES_REQUESTED** [review]

**Comment by fd0** [comment]
I agree that it's also valuable for the s3 backend. Hm. To be honest, I'd rather have a global option `--cacert` or so that is then used for all backends that use the `Transport`. The most straight-forward way to pass that option to the `Transport` would be to instanciate it in `global.go` and pass it as a parameter to all backends that need it. What do you think? Is that clear enough?

File: internal/backend/http_transport.go (hunk 1 above)
Comment (fawick) [reply internal/backend/http_transport.go]: Ooops, you're right, of course. I understand that you to write meant `!=`.

File: internal/backend/http_transport.go (hunk 2 above)
Comment (fawick) [reply internal/backend/http_transport.go]: Done.

**Comment by fawick** [comment]
I'm not against the idea to turn this into a global option. When I started out, I concentrated purely on the REST backend and did not think of it being of use to S3 then.

I want to raise one question, though: Should we be ready for the potential situation of the future in which restic instantiates more than one backend at runtime (e.g. repo migration, repo syncing, repo duplication)? It's easy to create a hypothetical scenario, in which the REST backend uses a custom CA for a self-signed certificate, whereas the the S3/GCS backends would still rely on Amazon's/Google's official certificates. Here it were necessary to be able to set the `--cacert` on a per-backend basis. Or even per repo (AWS S3 <-> self-hosted Minio)...

**Comment by fd0** [comment]
Let's worry about that use case when we get there :)

Another possible solution would be to allow more than one `--cacert` parameter and create a pool of allowed CA certificates (system and specified files)...

**Comment by fawick** [comment]
I'd agree that multiple `--cacert` are the best trade-off here, so I am going to implement that.

**Comment by fawick** [comment]
@fd0 Please take another look and let's discuss. I am not happy with the way I pass the filenames around, as it feels a bit clunky and reduces code readability IMHO. And I still need to come up with some tests for all that.

**Comment by fd0** [comment]
The code looks good already. I'd suggest we don't pass the filenames to `Open()` and `Create()`, but instead create the `http.Transport` in `global.go` and pass it as a parameter to `Open()` and `Create()` instead. This way we can add other (global) options like timeouts, and make this generic for all backends that use HTTP, this helps debugging also with other backends.

**Comment by fd0** [comment]
Do you need any more assistance? Or just not enough time? :)

**Comment by fawick** [comment]
WIP - do not review yet.

**Comment by fd0** [comment]
Looks finished, is it?

**Comment by fawick** [comment]
Yeah, expect for a test. But I can add that later.

**Review by fd0 — APPROVED** [review]

**Pull request merged**